evalml/pipelines/components/transformers/imputers/knn_imputer.py

```diff
@@ -0,0 +1,158 @@
+"""Component that imputes missing data according to a specified imputation strategy."""
+import numpy as np
+import pandas as pd
+import woodwork
+from sklearn.impute import KNNImputer as Sk_KNNImputer
+
+from evalml.pipelines.components.transformers import Transformer
+from evalml.utils import infer_feature_types
+from evalml.utils.gen_utils import is_categorical_actually_boolean
+
+
+class KNNImputer(Transformer):
+    """Imputes missing data using KNN according to a specified number of neighbors.  Natural language columns are ignored.
+
+    Args:
+        number_neighbors (int): Number of nearest neighbors for KNN to search for. Defaults to 0.
+        random_seed (int): Seed for the random number generator. Defaults to 0.
+
+    """
+
+    name = "KNN Imputer"
+
+    def __init__(self, number_neighbors=3, random_seed=0, **kwargs):
+        parameters = {"number_neighbors": number_neighbors}
+        parameters.update(kwargs)
+
+        imputer = Sk_KNNImputer(
+            n_neighbors=number_neighbors,
+            missing_values=np.nan,
+            **kwargs,
+        )
+        self._all_null_cols = None
+        super().__init__(
+            parameters=parameters,
+            component_obj=imputer,
+            random_seed=random_seed,
+        )
+
+    def _drop_natural_language_columns(self, X):
+        natural_language_columns = list(
+            X.ww.select(["NaturalLanguage"], return_schema=True).columns.keys(),
+        )
+        if natural_language_columns:
+            X = X.ww.copy()
+            X = X.ww.drop(columns=natural_language_columns)
+        return X, natural_language_columns
+
+    def _set_boolean_columns_to_categorical(self, X):
+        boolean_columns = list(
+            X.ww.select(["Boolean"], return_schema=True).columns.keys(),
+        )
+        if boolean_columns:
+            X = X.ww.copy()
+            X.ww.set_types({col: "Categorical" for col in boolean_columns})
+        return X
+
+    def fit(self, X, y=None):
+        """Fits imputer to data. 'None' values are converted to np.nan before imputation and are treated as the same.
+
+        Args:
+            X (pd.DataFrame or np.ndarray): the input training data of shape [n_samples, n_features]
+            y (pd.Series, optional): the target training data of length [n_samples]
+
+        Returns:
+            self
+
+        Raises:
+            ValueError: if the KNNImputer receives a dataframe with both Boolean and Categorical data.
+
+        """
+        X = infer_feature_types(X)
+
+        if set([lt.type_string for lt in X.ww.logical_types.values()]) == {
+            "boolean",
+            "categorical",
+        } and not all(
+            [
+                is_categorical_actually_boolean(X, col)
+                for col in X.ww.select("Categorical")
+            ],
+        ):
+            raise ValueError(
+                "KNNImputer cannot handle dataframes with both boolean and categorical features.  Use Imputer instead.",
```

Comment:
so can `KNNImputer` handle the case where there are only boolean and only categorical features? Why is this the case if we set boolean columns to categorical below?